pkg/upgrade/upgrade.go

```diff
@@ -0,0 +1,246 @@
+package upgrade
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/mitchellh/go-ps"
+	"github.com/opencontainers/selinux/go-selinux"
+	"github.com/rancher/k3s-upgrade/pkg/types"
+	"github.com/sirupsen/logrus"
+	"gopkg.in/yaml.v2"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+func fileExists(path string) bool {
+	_, err := os.Stat(path)
+	if os.IsNotExist(err) {
+		return false
+	}
+	return true
+}
+
+func GetProcessPID() (int, error) {
+	processes, err := ps.Processes()
+	if err != nil {
+		return 0, err
+	}
+	for _, proc := range processes {
+		if strings.HasSuffix(proc.Executable(), "k3s") {
+			if strings.Contains(proc.Executable(), "init") || strings.Contains(proc.Executable(), "channelserver") {
+				continue
+			}
+			return proc.Pid(), nil
+		}
+	}
+	return -1, fmt.Errorf("k3s process is not running")
+}
+
+func GetBinPath(pid int) (string, error) {
+	_, err := os.FindProcess(pid)
+	if err != nil {
+		return "", fmt.Errorf(" process is not found")
+	}
+	if pid == 1 {
+		return "/bin/k3s", nil
+	}
+	binPath, err := ioutil.ReadFile(path.Join("/host/proc", strconv.Itoa(pid), "cmdline"))
+	if fileExists(string(binPath)) {
+		return string(binPath), nil
+	}
+	return "", fmt.Errorf(" failed to fetch the k3s binary path from process %d", pid)
+}
+
+func TermProcess(pid int) error {
+	proc, err := os.FindProcess(pid)
+	if err != nil {
+		fmt.Errorf(" process is not found")
+	}
+	return proc.Signal(syscall.SIGTERM)
+}
+
+func MatchSelector(hostPath, configPath string, preservedEntries []string) error {
+	kubeconfig := os.Getenv("KUBECONFIG")
+	conf, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		return err
+	}
+	nodeClient := kubernetes.NewForConfigOrDie(conf).CoreV1().Nodes()
+	hostname := os.Getenv("SYSTEM_UPGRADE_NODE_NAME")
```

Comment:
Prefer to call this `nodeName` instead of `hostname`. This is kind of a nit but it is important to not confuse the two. A kubelet's node name can be different than the hostname it is running on.